_posts/2015-06-02-anatomy-of-service-objects-in-rails.md

```diff
@@ -0,0 +1,333 @@
+
+What you *should not* be passing to your service object's methods are other service objects.  We saw that `ReturnProcessor` will charge customers for unreturned items using an instance of `CheckoutService`.  
+
+Passing it in requires all callers to deal with how to create a `CheckoutService` properly as well as exposes them to the internal implementation of `process!`, thus making it harder to change:
+
+```ruby
+# Bad
+ReturnProcessor.new.process!(the_return,user,CheckoutService.new)
+```
+
+If `process!` simply accepts the data its operating on and/or needs to read, it makes more sense.  The caller is going to have the return being processed and the user processing it.
+
+```ruby
+# Good
+ReturnProcessor.new.process!(the_return,user)
+```
+
+Passing in dependent service objects puts undue burden on the caller, requiring code duplication and needless coupling to the service object's implementation.
+
+We'll see in a second how the service object should get access to dependent service objects.  But first, let’s see what sorts of data our methods should be returning.
+
+## Methods should return rich result objects, not booleans
+
+Service object methods typically have three possible outcomes:
+
+* The requested action succeeded
+* It failed, but in an expected way
+* An exceptional condition occurred
+
+Distinguishing failure from exceptions can be subtle, but it's important.  A failure is something the caller is expected to deal with. The distinction between success and failure is also potentially subtle, as both represent “valid things that could happen as a result of this call”.  Those valid things could require additional context to interpret.  
+
+For example, a successful return might require that some items be set aside for donation, and others returned to inventory.  A failed return might need to indicate which items were accounted for incorrectly and a suggestion as to how the user can resolve it.  A simple `true` or `false` doesn’t communicate that information, nor could it ever.
+
+This means that the common pattern of returning `true` or `false` should be avoided. Neither `true` nor `false` can contain any sort of context. They are also not extensible in any way.  If a simple `true` is fine today, but you later need to return more information, you’ve got a much bigger re-design on your hands than if you’d used a richer object initially.
+
+For this reason, I prefer returning result objects that have a more fluent API.  Creating these objects is trivial with libraries like [immutable-struct][is]<a name="back-1"></a><sup><a href="#1">1</a></sup>:
+
+```ruby
+class ReturnProcessor
+  Result = ImmutableStruct.new(:return_processed?, 
+                               :error_messages)
+end
+```
+
+Notice that instead a method like `success?`, I’ve used an explicit name for what happen—`return_processed?`.  Users of `ReturnProcessor` will then benefit from clear, intention-revealing code:
```

Comment:
"instead a method" should be "instead of a method"
"happen" should be "happened"